File: LeetCode_Python/List&LinkList/160.intersection-of-two-linked-lists.py
# @before-stub-for-debug-begin
from python3problem160 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=160 lang=python3
#
# [160] Intersection of Two Linked Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        p1 = headA 
        p2 = headB 
        while(p1 is not p2):
            if (p1 is None):
                p1 = headB
            else:
                p1 = p1.next

            if (p2 is None):
                p2 = headA
            else:
                p2 = p2.next

        return p1
        

  # 2个相同速度的指针 p1, p2
  # 设定2个链表的共同部分为k
  # 设定a链表的前置 skA, 设定b链表的前置 skB
  # p1: skA -> a - > skB -> c1
  # p2: skB -> a - > skA -> c1
  # p1 与 p2 必在c1相遇
  # 如果2个链表没有相交 (a为空）他们会在None相遇

# @lc code=end

File: LeetCode_Go/List&LinkedList/160.intersection-of-two-linked-lists.go
/*
 * @lc app=leetcode id=160 lang=golang
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// Given two acyclic lists, return the node at which the two lists intersect. If the two linked lists have no intersection at all, return null.

// two pointers, same speed, each start from the head of one list, finishing traverse the list, and then start to traverse the other list, (at the end both pointer will traverse each nodes in both lists). If they both reached nil without meeting, there is no intersection. Otherwise, they will meet at the same node, which is the intersection node.

// O(m+n) time, O(1) space

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	pointA, pointB := headA, headB
	for pointA != nil || pointB != nil {
		// fmt.Printf("pointA: %v, pointB: %v\n", pointA.Val, pointB.Val)
		if pointA == pointB { //found intersection
			return pointA
		}
		pointA = pointA.Next
		pointB = pointB.Next
		if pointA == nil && pointB != nil {
			pointA = headB
		}
		if pointB == nil && pointA != nil {
			pointB = headA
		}
	}
	return nil
}

// @lc code=end

